File: core/src/main/scala/global/namespace/neuron/di/scala/Neuron.scala
/*
 * Copyright © 2016 Schlichtherle IT Services
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package global.namespace.neuron.di.scala

import global.namespace.neuron.di.java.{Neuron => jNeuron}

import scala.annotation.{StaticAnnotation, compileTimeOnly}
import scala.language.experimental.macros
import scala.reflect.macros.whitebox

@compileTimeOnly("Please add the Macro Paradise plugin to the Scala compiler to expand this macro annotation.")
class Neuron extends StaticAnnotation {

  def macroTransform(annottees: Any*): Any = macro Neuron.impl
}

private object Neuron {

  def impl(c: whitebox.Context)(annottees: c.Tree*): c.Tree = {
    import c.universe._
    import Flag._

    implicit class FlagOps(left: FlagSet) {

      // As of Scala 2.11.8:
      private val allFlags = Set(
        ABSOVERRIDE,
        ABSTRACT,
        ARTIFACT,
        BYNAMEPARAM,
        CASE,
        CASEACCESSOR,
        CONTRAVARIANT,
        COVARIANT,
        DEFAULTINIT,
        DEFAULTPARAM,
        DEFERRED,
        ENUM,
        FINAL,
        IMPLICIT,
        INTERFACE,
        LAZY,
        LOCAL,
        MACRO,
        MUTABLE,
        OVERRIDE,
        PARAM,
        PARAMACCESSOR,
        PRESUPER,
        PRIVATE,
        PROTECTED,
        SEALED,
        STABLE,
        SYNTHETIC,
        TRAIT
      )

      def &~(right: FlagSet): FlagSet = {
        var result = NoFlags
        for (flag <- allFlags) {
          if (flag == (left | flag) && flag != (right | flag)) {
            result |= flag
          }
        }
        result
      }
    }

    val inputs = annottees.toList
    val outputs: List[Tree] = inputs match {
      case ClassDef(mods @ Modifiers(flags, privateWithin, annotations), typeName @ TypeName(name), tparams, impl) :: rest =>
        val neuron = Apply(Select(New(Select(Select(Select(Select(Select(Select(Ident(termNames.ROOTPKG), TermName("global")), TermName("namespace")), TermName("neuron")), TermName("di")), TermName("java")), TypeName("Neuron"))), termNames.CONSTRUCTOR), Nil)
        ClassDef(mods.mapAnnotations(neuron :: _), typeName, tparams, impl) :: {
          if (mods.hasFlag(TRAIT) && !mods.hasFlag(INTERFACE)) {
            val implDef = ClassDef(Modifiers(flags &~ (TRAIT | DEFAULTPARAM) | ABSTRACT | SYNTHETIC, privateWithin, neuron :: annotations), TypeName("$shim"), Nil, Template(List(Ident(typeName)), noSelfType, List(DefDef(Modifiers(), termNames.CONSTRUCTOR, Nil, List(Nil), TypeTree(), Block(List(pendingSuperCall), Literal(Constant(())))))))
            rest match {
              case ModuleDef(mods, name, Template(parents, self, body)) :: rest =>
                ModuleDef(mods, name, Template(parents, self, implDef :: body)) :: rest
              case rest =>
                ModuleDef(Modifiers(flags &~ (ABSTRACT | TRAIT | DEFAULTPARAM) | SYNTHETIC, privateWithin, Nil), TermName(name), Template(Nil, noSelfType, List(implDef, DefDef(Modifiers(), termNames.CONSTRUCTOR, List(), List(List()), TypeTree(), Block(List(pendingSuperCall), Literal(Constant(()))))))) :: rest
            }
          } else {
            rest
          }
        }
      case _ =>
        inputs
    }
    q"..$outputs"
  }
}

File: core/src/main/java/global/namespace/neuron/di/java/Neuron.java
/*
 * Copyright © 2016 - 2019 Schlichtherle IT Services
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package global.namespace.neuron.di.java;

import java.lang.annotation.*;

/**
 * Indicates that the annotated class or interface is a <i>neuron type</i>.
 * A neuron type is supposed to have <i>synapse methods</i> to make contact with its dependencies.
 * All abstract, parameterless methods in a neuron type are synapse methods - there is no need to declare them.
 * These methods get automatically implemented so that they return their respective dependencies.
 * Thus, all dependencies of a neuron are resolved lazily plus they have a name and a type.
 * <p>
 * If this annotation is present on a class, it gets inherited by any subclass.
 * In compliance with the Java Language Specification, this does not apply to interfaces.
 * <p>
 * This annotation can also be applied to custom annotation types, in which case any class or interface annotated with
 * these is treated as if it's directly annotated with this annotation.
 * For example, the following annotation caches all synapse values using a thread-safe strategy:
 * <pre><code>
 *     import global.namespace.neuron.di.java.*;
 *     import java.lang.annotation.*;
 *
 *     &#64;Neuron(cachingStrategy = CachingStrategy.THREAD_SAFE)
 *     &#64;Inherited
 *     &#64;Documented
 *     &#64;Retention(RetentionPolicy.RUNTIME)
 *     &#64;Target(ElementType.TYPE)
 *     public &#64;interface ThreadSafeCachingNeuron {
 *     }
 * </code></pre>
 *
 * @see <a href="https://en.wikipedia.org/wiki/Neuron">Neuron on the English Wikipedia</a>
 */
@Inherited
@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
public @interface Neuron {

    /**
     * Configures the strategy for caching the return value of synapse methods which are not annotated with
     * {@link Caching}.
     */
    CachingStrategy cachingStrategy() default CachingStrategy.DISABLED;
}
